builtin/providers/aws/resource_aws_dms_endpoint.go

```diff
@@ -0,0 +1,334 @@
+package aws
+
+import (
+	"log"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/private/waiter"
+	dms "github.com/aws/aws-sdk-go/service/databasemigrationservice"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceAwsDmsEndpoint() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsDmsEndpointCreate,
+		Read:   resourceAwsDmsEndpointRead,
+		Update: resourceAwsDmsEndpointUpdate,
+		Delete: resourceAwsDmsEndpointDelete,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"certificate_arn": {
+				Type:         schema.TypeString,
+				Computed:     true,
+				Optional:     true,
+				ValidateFunc: validateArn,
+			},
+			"database_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+				Optional: true,
+			},
+			"endpoint_arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"endpoint_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"endpoint_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"source",
+					"target",
+				}, false),
+			},
+			"engine_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				// Valid Values: mysql, oracle, postgres, mariadb, aurora, redshift, sybase, sqlserver
```

Comment:
I added the validate function. This probably won't change very often. 

For something like replication_instance_type where new instance types may be added often over the course of the year, should we validate those too?